src/Stratis.Bitcoin/EventBus/InMemoryEventBus.cs

```diff
@@ -0,0 +1,89 @@
+using System.Linq;
+using System.Text;
+
+namespace Stratis.Bitcoin.EventBus
+{
+    public class InMemoryEventBus : IEventBus
+    {
+        /// <summary>
+        /// The subscriptions stored by EventType
+        /// </summary>
+        private readonly Dictionary<Type, List<ISubscription>> subscriptions;
+
+        /// <summary>
+        /// The subscriptions lock to prevent race condition during publishing
+        /// </summary>
+        private static readonly object SubscriptionsLock = new object();
+
+        public InMemoryEventBus()
+        {
+            this.subscriptions = new Dictionary<Type, List<ISubscription>>();
+        }
+
+        /// <inheritdoc />
+        public SubscriptionToken Subscribe<TEventBase>(Action<TEventBase> action) where TEventBase : EventBase
+        {
+            if (action == null)
+                throw new ArgumentNullException(nameof(action));
+
+            lock (SubscriptionsLock)
+            {
+                if (!this.subscriptions.ContainsKey(typeof(TEventBase)))
+                    this.subscriptions.Add(typeof(TEventBase), new List<ISubscription>());
+
+                var token = new SubscriptionToken(this, typeof(TEventBase));
+                this.subscriptions[typeof(TEventBase)].Add(new Subscription<TEventBase>(action, token));
+                return token;
+            }
+        }
+
+        /// <inheritdoc />
+        public void Unsubscribe(SubscriptionToken token)
+        {
+            if (token == null)
+                throw new ArgumentNullException(nameof(token));
+
+            lock (SubscriptionsLock)
+            {
+                if (this.subscriptions.ContainsKey(token.EventType))
+                {
+                    var allSubscriptions = this.subscriptions[token.EventType];
+
+                    var subscriptionToRemove = allSubscriptions.FirstOrDefault(sub => sub.SubscriptionToken.Token == token.Token);
+                    if (subscriptionToRemove != null)
+                        this.subscriptions[token.EventType].Remove(subscriptionToRemove);
+                }
+            }
+        }
+
+        /// <inheritdoc />
+        public void Publish<TEventBase>(TEventBase eventItem) where TEventBase : EventBase
+        {
+            if (eventItem == null)
+                throw new ArgumentNullException(nameof(eventItem));
+
+            List<ISubscription> allSubscriptions = new List<ISubscription>();
+            lock (SubscriptionsLock)
+            {
+                if (this.subscriptions.ContainsKey(typeof(TEventBase)))
+                    allSubscriptions = this.subscriptions[typeof(TEventBase)].ToList();
+            }
+
+            for (var index = 0; index < allSubscriptions.Count; index++)
+            {
+                var subscription = allSubscriptions[index];
+                try
+                {
+                    subscription.Publish(eventItem);
+                }
+                catch (Exception ex)
+                {
```

Comment:
This should probably throw to the caller for now (same functionality as before)

Reply:
Probably implementing the SubscriptionErrorHandler is a best idea, but to have compatibility I can throw the exception for the moment

Reply:
`SubscriptionErrorHandler ` yeah that sounds right, with a default one that will thorw

Reply:
done